src/agent/windows/WMICommunication.cpp

```diff
@@ -0,0 +1,215 @@
+        {
+            throw std::exception("Could not connect. Error code = 0x");
+        }
+
+        hres = CoSetProxyBlanket(
+            m_services,                    // Indicates the proxy to set
+            RPC_C_AUTHN_WINNT,           // RPC_C_AUTHN_xxx
+            RPC_C_AUTHZ_NONE,            // RPC_C_AUTHZ_xxx
+            NULL,                        // Server principal name 
+            RPC_C_AUTHN_LEVEL_CALL,      // RPC_C_AUTHN_LEVEL_xxx 
+            RPC_C_IMP_LEVEL_IMPERSONATE, // RPC_C_IMP_LEVEL_xxx
+            NULL,                        // client identity
+            EOAC_NONE                    // proxy capabilities 
+        );
+
+        if (FAILED(hres))
+        {
+            throw std::exception("Could not set proxy blanket. Error code = 0x");
+        }
+
+        return true;
+    }
+    catch (...)
+    {
+        CoUninitialize();
+        if (m_initialLocatorToWMI != NULL)
+        {
+            m_initialLocatorToWMI->Release();
+        }
+
+        if (m_services != NULL)
+        {
+            m_services->Release();
+        }
+
+        return false;
+    }
+}
+
+bool WMICommunication::GetSMARTDataViaWMI()
+{
+    try {
+        HRESULT hres = m_services->ExecQuery(
+            bstr_t("WQL"),
+            bstr_t("SELECT * FROM MSStorageDriver_FailurePredictData"),
+            WBEM_FLAG_FORWARD_ONLY | WBEM_FLAG_RETURN_IMMEDIATELY,
+            NULL,
+            &m_pEnumerator);
+
+        if (FAILED(hres))
+        {
+            throw std::exception("Query for operating system name failed. Error code = 0x");
+        }
+
+        IWbemClassObject* pclsObj = NULL;
+        ULONG uReturn = 0;
+
+        while (m_pEnumerator)
+        {
+            HRESULT hr = m_pEnumerator->Next(WBEM_INFINITE, 1,
+                &pclsObj, &uReturn);
+
+            if (0 == uReturn)
+            {
+                break;
+            }
+
+            VARIANT vtProp;
+            hr = pclsObj->Get(L"VendorSpecific", 0, &vtProp, 0, 0);
+
+            if (V_ISARRAY(&vtProp))
+            {
+                LPSAFEARRAY pSafeArray = V_ARRAY(&vtProp);
+
+                VARTYPE itemType;
+                if (SUCCEEDED(SafeArrayGetVartype(pSafeArray, &itemType)))
+                {
+                    if (itemType == VT_UI1)
+                    {
+                        if (SafeArrayGetDim(pSafeArray) == 1)
+                        {
+                            LONG lBound;
+                            LONG uBound;
+
+                            if (SUCCEEDED(SafeArrayGetLBound(pSafeArray, 1, &lBound)) && SUCCEEDED(SafeArrayGetUBound(pSafeArray, 1, &uBound)))
+                            {
+                                const LONG itemCount = uBound - lBound + 1;
+
+                                std::vector<BYTE> dataFromArray(itemCount);
+
+                                BYTE* safearrayData;
+                                hr = SafeArrayAccessData(pSafeArray, reinterpret_cast<LPVOID*>(&safearrayData));
```

Comment:
I'm 60% sure this reinterpret cast is not needed, is it? if castng is needed then less hardcore static_cast should work.

Reply:
Error	C2440	'static_cast': cannot convert from 'BYTE **' to 'LPVOID *'	
So we have here this 40% :)